src/array.cr

```diff
@@ -1046,8 +1046,8 @@ class Array(T)
   end
 
   # Optimized version of `Enumerable#map`.
-  def map(& : T -> U) forall U
-    Array(U).new(size) { |i| yield @buffer[i] }
+  def map(& : T -> _)
+    Array.new(size) { |i| yield @buffer[i] }
```

Comment:
I fear this would be a breaking change. The return type isn't `Array(T)` anymore. Instead, its inferred from the output type of the block, essentially `Array(typeof(yield @buffer[0]))`. This is an inacceptable mutation of the API contract.

Reply:
You're totally right, I slipped this one in by mistake

Reply:
That's `#map`, not `#map!`, so I think it's fine.

Reply:
Of course it returns a new instance. But the return type used to be the same generic type instance, with this change it doesn't anymore.
Maybe that could be considered a valid behaviour as well, but I doubt it. You can get type errors when you assign `@foo = @foo.map { ... }` and the block outputs only a subset of the generic argument types.
In any case, such a change is a major breaking change. It's not reasonable to tolerate that just to add a workaround for an edge case bug.

Reply:
I don't understand. `[1, 'a'].map(&.to_s)` and `[1, "aaa"].map(&.to_s)` are already valid behaviour. The fact the assignment to `@foo` fails isn't due to the free variable in question.

Reply:
Oh, I think I mixed up `T` and `U`. The generic type argument would already be the output type of the block and not identical to the type of `self`.

Sorry. Yeah, I suppose this should be fine.

Reply:
I think in recursive calls this might break. Safe bet is to keep it like it was before.